content/blog/unsafe-string-interning-in-go.md

```diff
@@ -0,0 +1,345 @@
++++
+
+* * * * * DRAFT * * * * *
+
+For a few months, I've been collaborating on the
+[`inet.af/netaddr`](https://inet.af/netaddr) package for Go which introduces a
+new `netaddr.IP` type as an alternative to the standard library's
+[`net.IP`](https://golang.org/pkg/net/#IP) type. There are [a variety of reasons
+for this](https://pkg.go.dev/inet.af/netaddr#readme-motivation), but those are
+not the focus of today's blog.
+
+The folks over at [Tailscale](https://tailscale.com/) have been working on
+improving the performance of the package in various ways, including shrinking
+the size of types and trying out faster internal IP representations.
+Specifically, one topic that came up was [how to shrink `netaddr.IP` from 32
+bytes to 24 bytes](https://github.com/inetaf/netaddr/issues/), the same size
+as a Go slice header.
+
+Because a `netaddr.IP` can also have an associated zone for IPv6 link-local
+addresses, the type previously contained a Go string, which occupies 16 bytes.
+This is a good use case for [string
+interning](https://en.wikipedia.org/wiki/String_interning), a technique which
+can be used to reduce the number of string allocations and shrink the type via a
+pointer field. It is highly likely that a small number of zone identifiers will
+be used repeatedly throughout a program, so string interning seems like a great
+fit for this problem.
+
+The result of this work is the package
+[`go4.org/intern`](https://go4.org/intern) which uses some pretty neat `unsafe`
+tricks to implement efficient string interning using weak references and Go
+finalizers. We'll start by showing off the safe implementation and gradually
+introduce the concepts needed to understand the unsafe one as well.
+
+All credit for the ideas and implementation of `go4.org/intern` goes to [Dave
+Anderson](https://twitter.com/dave_universetf), [Josh Bleecher
+Snyder](https://twitter.com/commaok), and [Brad
+Fitzpatrick](https://twitter.com/bradfitz). I simply watched the discussions
+unfold on the issue tracker and thought that the code and the ideas behind it
+were too interesting to not write up and share!
+
+## Safe string interning in Go
+
+The basic idea of string interning is to retain a global pool of strings which
+can be created or fetched on the fly, reducing the number of repeated
+allocations for values which are identical. In its most basic form, you could
+implement `go4.org/intern` (with a caveat!) by writing the following:
+
+```go
+package intern
+
+// A Value pointer is the handle to an underlying comparable value.
+// See func Get for how Value pointers may be used.
+type Value struct {
+	// Enforce that Values cannot be used directly for equality comparisons.
+	// We'll explain why a bit later.
+	_      [0]func()
+	cmpVal interface{}
+}
+
+// Get returns the comparable value passed to the Get func
+// that returned v.
+func (v *Value) Get() interface{} { return v.cmpVal }
```

Comment:
go4.org/intern doesn't just intern strings; it interns any comparable value. That's why we use `interface{}` in the API. Given that you talk only about strings in the intro, might want to call that out before sharing this code.

Also, you might want to call out that the reason we return `*Value` from Get is so that the handles you get from package intern are only one word long, despite strong a possibly massive amount of data under the hood.